File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.java
package com.virtuslab.gitmachete.frontend.actions.dialogs;

import javax.swing.Action;
import javax.swing.JComponent;

import com.intellij.openapi.ui.DialogWrapper;
import com.intellij.ui.ScrollPaneFactory;
import com.intellij.ui.table.JBTable;
import com.intellij.util.Consumer;
import com.intellij.util.ui.JBDimension;
import com.intellij.util.ui.JBUI;
import org.checkerframework.checker.guieffect.qual.UI;
import org.checkerframework.checker.nullness.qual.Nullable;

import com.virtuslab.binding.RuntimeBinding;
import com.virtuslab.gitmachete.backend.api.IGitMacheteRepositorySnapshot;
import com.virtuslab.gitmachete.frontend.ui.api.table.ISimpleGraphTableProvider;

@UI
public final class GraphTableDialog extends DialogWrapper {
  private final JBTable table;
  private final @Nullable IGitMacheteRepositorySnapshot repositorySnapshot;
  private final int windowWidth;
  private final int windowHeight;
  private final @Nullable Consumer<IGitMacheteRepositorySnapshot> okAction;
  private final boolean cancelButtonVisible;

  public static GraphTableDialog of(IGitMacheteRepositorySnapshot gitMacheteRepositorySnapshot, String windowTitle,
      @Nullable String emptyTableText, @Nullable Consumer<IGitMacheteRepositorySnapshot> okAction, String okButtonText,
      boolean cancelButtonVisible) {
    return of(gitMacheteRepositorySnapshot, windowTitle, /* windowWidth */ 800, /* windowHeight */ 500, emptyTableText,
        okAction, okButtonText, cancelButtonVisible);
  }

  public static GraphTableDialog of(IGitMacheteRepositorySnapshot gitMacheteRepositorySnapshot, String windowTitle,
      int windowWidth, int windowHeight, @Nullable String emptyTableText,
      @Nullable Consumer<IGitMacheteRepositorySnapshot> okAction, String okButtonText, boolean cancelButtonVisible) {
    var table = RuntimeBinding.instantiateSoleImplementingClass(ISimpleGraphTableProvider.class)
        .deriveInstance(gitMacheteRepositorySnapshot);
    table.setTextForEmptyTable(emptyTableText != null ? emptyTableText : "");
    return new GraphTableDialog(table, gitMacheteRepositorySnapshot, windowTitle, windowWidth, windowHeight, okAction,
        okButtonText, cancelButtonVisible);
  }

  public static GraphTableDialog ofDemoRepository() {
    var table = RuntimeBinding.instantiateSoleImplementingClass(ISimpleGraphTableProvider.class).deriveDemoInstance();
    return new GraphTableDialog(table, /* repositorySnapshot */ null, /* windowTitle */ "Git Machete Help",
        /* windowWidth */800, /* windowHeight */ 250, /* okAction */ null, /* okButtonText */ "Close",
        /* cancelButtonVisible */ false);
  }

  private GraphTableDialog(JBTable table, @Nullable IGitMacheteRepositorySnapshot repositorySnapshot, String windowTitle,
      int windowWidth, int windowHeight, @Nullable Consumer<IGitMacheteRepositorySnapshot> okAction, String okButtonText,
      boolean cancelButtonVisible) {
    super(/* canBeParent */ false);

    this.table = table;
    this.repositorySnapshot = repositorySnapshot;
    this.windowWidth = windowWidth;
    this.windowHeight = windowHeight;
    this.okAction = okAction;
    this.cancelButtonVisible = cancelButtonVisible;

    // Note: since the class is final, `this` is already @Initialized at this point.
    init();
    setTitle(windowTitle);
    setOKButtonText(okButtonText);
  }

  @Override
  protected Action[] createActions() {
    return cancelButtonVisible
        ? new Action[]{getOKAction(), getCancelAction()}
        : new Action[]{getOKAction()};
  }

  @Override
  protected JComponent createCenterPanel() {
    var panel = JBUI.Panels.simplePanel(/* hgap */ 0, /* vgap */ 2);
    var scrollPane = ScrollPaneFactory.createScrollPane(table);
    panel.addToCenter(scrollPane);
    panel.setPreferredSize(new JBDimension(windowWidth, windowHeight));
    return panel;
  }

  @Override
  protected void doOKAction() {
    if (getOKAction().isEnabled()) {
      if (okAction != null && repositorySnapshot != null) {
        okAction.consume(repositorySnapshot);
      }
      close(OK_EXIT_CODE);
    }
  }
}

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GraphTableDialog.kt
package com.virtuslab.gitmachete.frontend.actions.dialogs

import com.intellij.openapi.ui.DialogWrapper
import com.intellij.ui.ScrollPaneFactory
import com.intellij.ui.table.JBTable
import com.intellij.util.ui.JBDimension
import com.intellij.util.ui.JBUI
import com.virtuslab.binding.RuntimeBinding
import com.virtuslab.gitmachete.backend.api.IGitMacheteRepositorySnapshot
import com.virtuslab.gitmachete.frontend.ui.api.table.ISimpleGraphTableProvider
import java.awt.event.ActionEvent
import java.util.function.Consumer
import javax.swing.AbstractAction

class GraphTableDialog
    private constructor(
        private val table: JBTable,
        private val repositorySnapshot: IGitMacheteRepositorySnapshot?,
        private val dimension: JBDimension,
        private val okAction: Consumer<IGitMacheteRepositorySnapshot>?,
        private val editAction: Runnable?,
        private val cancelButtonVisible: Boolean,
        private val windowTitle: String,
        private val okButtonText: String
    ) : DialogWrapper(/* canBeParent */ false) {

  init {
    super.init()
    title = windowTitle
    setOKButtonText(okButtonText)
  }

  companion object {
    // @JvmOverloads annotation instructs the Kotlin compiler to generate overloads
    // for this function that substitute default parameter values (dimension in our case).
    @JvmOverloads
    fun of(
        gitMacheteRepositorySnapshot: IGitMacheteRepositorySnapshot,
        windowTitle: String,
        dimension: JBDimension = JBDimension(/* width */ 800, /* height */ 500),
        emptyTableText: String?,
        okAction: Consumer<IGitMacheteRepositorySnapshot>?,
        editAction: Runnable?,
        okButtonText: String,
        cancelButtonVisible: Boolean,
        hasBranchActionToolTips: Boolean
    ) =
        RuntimeBinding.instantiateSoleImplementingClass(ISimpleGraphTableProvider::class.java)
            .deriveInstance(
                gitMacheteRepositorySnapshot,
                /* isListingCommitsEnabled */ false,
                hasBranchActionToolTips)
            .apply { setTextForEmptyTable(emptyTableText ?: "") }
            .let {
              GraphTableDialog(
                  /* table */ it,
                  gitMacheteRepositorySnapshot,
                  dimension,
                  okAction,
                  editAction,
                  cancelButtonVisible,
                  windowTitle,
                  okButtonText)
            }

    fun ofDemoRepository() =
        RuntimeBinding.instantiateSoleImplementingClass(ISimpleGraphTableProvider::class.java)
            .deriveDemoInstance()
            .let {
          GraphTableDialog(
              table = it,
              repositorySnapshot = null,
              dimension = JBDimension(/* width */ 800, /* height */ 250),
              okAction = null,
              editAction = null,
              cancelButtonVisible = false,
              windowTitle = "Git Machete Help",
              okButtonText = "Close")
        }
  }

  override fun createActions() =
      if (cancelButtonVisible)
          arrayOf(getSaveAndEditAction(), getOKAction(), cancelAction)
              .filterNotNull()
              .toTypedArray()
      else arrayOf(getOKAction())

  private fun getSaveAndEditAction() =
      if (editAction == null) null
      else
          object : AbstractAction("Save && Edit") {
            override fun actionPerformed(e: ActionEvent?) {
              doOKAction()
              editAction.run()
            }
          }

  override fun createCenterPanel() =
      JBUI.Panels.simplePanel(/* hgap */ 0, /* vgap */ 2).apply {
        addToCenter(ScrollPaneFactory.createScrollPane(this@GraphTableDialog.table))
        preferredSize = dimension
      }

  @Override
  override fun doOKAction() {
    repositorySnapshot?.let { okAction?.accept(it) }
    close(OK_EXIT_CODE)
  }
}
